Selector thomas

dupriezt opened the pull request:
New class: DebugRequest. Places in the system that want to open debuggers create one, fill it with the info they have (exception or context, label, process if not default...) and send #submit to it.
New class: DebuggerSystem. Singleton. Handles DebugRequest by asking what the default UIManager wants to to do about it. Typically, the UI Manager will call back DebuggerSystem class>>#openDebuggerOnRequest. This method does some setup (that was previously done in DebuggerSelector) like performing the preDebug actions, spawning a new UI process if necessary... Then it uses DebuggerSelector as a strategy to select the best debugger, open it... as before.

Diagram with the new pipeline:
![2020-06-22_ProposedActionSequenceForOpeningDebuggersOnException](https://user-images.githubusercontent.com/32486709/85316156-92ebce00-b4bc-11ea-9153-7290e8ee2bc3.png)

Notes:
- Some of the responsibilities of DebuggerSelector have been moved to DebuggerSystem. But DebuggerSelector is still tied to many tests. For now, I created a class DebuggerSelector_ with the required methods of DebuggerSelector (and without the rest), and DebuggerSystem calls this new class. A future task is to prune DebuggerSelector of the now unused methods, adjust the tests, and replace DebuggerSelector_ with it.
- It works (executing 'self halt' in a playground opens a StDebugger). But there is a bug: closing a debugger window freezes the image. I don't know where it comes from. A future task is to find its cause and fix it.
- There are tests for DebuggerSystem
- The strategy design pattern in DebuggerSystem, for the strategy to select and open a debugger is not fully implemented, because we only have 1 strategy for it for now (DebuggerSelector). But the code is setup so that the strategy pattern can be implemented quickly

Commit:
Give StDebugger class a rank. Otherwise the initialization fails

Commit:
removing broken code in StDebugger>>#initialize

StevenCostiou committed:
Missing warning redirection call to DebuggerSelector

StevenCostiou committed:
Fixing post-mortem bug because debug sessions were initialized with the wrong signaler context

Commit:
Rework of the Exception -> Debugger opening chain.
Still needs polish.

Commit:
Fixed missing method

Commit:
DebugRequest no longer compute their preDebug actions when they get submitted. DebuggerSystem tells them to compute their preDebugActions before selecting a debugger to open them.
Tests for DebuggerSystem.

Commit:
Comments for DebuggerSystem tests.
Test for preDebugActions check that they are executed exactly once

StevenCostiou commented on src/NewTools-DebuggerSelector/DebuggerSystem.class.st:
```diff
@@ -0,0 +1,98 @@
+	^ self default openDebuggerOnRequest: aDebugRequest
+]
+
+{ #category : #'default values' }
+DebuggerSystem >> defaultDebuggerSelectionStrategy [
+	^ DebuggerSelector_
+]
+
```
À bouger coté classe

StevenCostiou replied:
Il faut aussi un test

StevenCostiou commented on src/NewTools-DebuggerSelector/DebuggerSystem.class.st:
```diff
@@ -0,0 +1,98 @@
+	^ default
+]
+
+{ #category : #'handle debug requests' }
+DebuggerSystem class >> handleDebugRequest: aDebugRequest [
+	"Forwards call to singleton instance"
+	^ self default handleDebugRequest: aDebugRequest
+]
+
+{ #category : #'handle debug requests' }
+DebuggerSystem class >> handleWarningDebugRequest: aWarningDebugRequest [
+	"Forwards call to singleton instance"
+	^ self default handleWarningDebugRequest: aWarningDebugRequest
+]
+
+{ #category : #'open debugger' }
+DebuggerSystem class >> openDebuggerOnRequest: aDebugRequest [
+	"Forwards call to singleton instance"
+	^ self default openDebuggerOnRequest: aDebugRequest
+]
```
Je n'ai pas vu de tests sur cette API

StevenCostiou reviewed (changes requested):
- faire sauter le singleton
- faire une passe sur le code (retirer les _ de tes noms de variables notamment)
- tester debuggerrequest et faire une PR séparée (le test guide et documente)
- tester debugger-selector
- tester debugger-system

dupriezt commented on src/NewTools-DebuggerSelector/DebugRequest.class.st:
```diff
@@ -0,0 +1,150 @@
+
+{ #category : #helper }
+DebugRequest >> nullExceptionFor: aContext named: aString [
+	| exception_ |
+	exception_ := DebuggerSelectorNullException
+		fromSignallerContext: aContext.
+	exception_ messageText: aString.
+	^ exception_
```
Il y a déjà une variable d'instance qui s'appelle "exception"

estebanlm replied:
then rename it as something more detailed: 
```Smalltalk
| nullException |
... etc
```
or you can cascade, you do not even need a variable declaration there.

dupriezt replied:
ok, thanks

dupriezt commented on src/NewTools-DebuggerSelector/DebugRequest.class.st:
```diff
@@ -0,0 +1,150 @@
+]
+
+{ #category : #getter }
+DebugRequest >> exception [
+	"If an exception was provided, return it, otherwise create a dummy exception using the provided context and label (if no label was provided, use the result of 'self defaultLabel_' "
+	^ exception ifNil: [ self nullExceptionFor: context named: (label ifNil: [ self defaultLabel_ ]) ]
+]
+
```
C'est expliqué dans DebugRequest>>#defaultLabel_

StevenCostiou replied:
non mais c'est interdit il n'y a pas d'explications à avoir :)

dupriezt commented on src/NewTools-DebuggerSelector/DebugRequest.class.st:
```diff
@@ -0,0 +1,150 @@
+	debugSession ifNil: [ debugSession := self process newDebugSessionNamed: self exception description startedAt: self exception signalerContext ].
+	^ debugSession.
+]
+
+{ #category : #'default values' }
+DebugRequest >> defaultLabel_ [
+	"The underscore is to prevent conflicting with Object>>#defaultLabel, which is a morphic method"
+	^ 'No label was provided'
+]
```
C'est expliqué dans le commentaire de la méthode

estebanlm replied:
is still a bad and not allowed name. 
You can always `detail` your names. Also here we do not care because your DebugRequest class is not in the morphic hierarchy. You can name it `defaultLabel` without danger to have a collision (also not wanting to repeat names when the meaning is the same is incorrect)

dupriezt replied:
ok, I'll do that. Thanks

dupriezt commented on src/NewTools-DebuggerSelector/DebuggerSystem.class.st:
```diff
@@ -0,0 +1,98 @@
+I am a singleton. Use my class side. My class side methods are redirects to the instance methods of my singleton method.
+I handle DebugRequests by asking the default UI Manager what to do about them. Typically, the UI manager will decide to open a debugger on the debugRequest, and call my #openDebuggerOnRequest: method to do so.
+"
+Class {
+	#name : #DebuggerSystem,
+	#superclass : #Object,
+	#classVars : [
+		'default'
+	],
+	#category : #'NewTools-DebuggerSelector-Core'
+}
+
```
Parce que ça veut dire créer une instance à chaque appel de l'API, alors que toutes ces instances seront identiques vu qu'elles n'ont pas d'état.

StevenCostiou replied:
Yes :)

dupriezt commented on src/NewTools-DebuggerSelector/DebuggerSystem.class.st:
```diff
@@ -0,0 +1,98 @@
+]
+
+{ #category : #helpers }
+DebuggerSystem >> spawnNewUIProcessIfNecessary: aDebugRequest [
+	"If @aDebugRequest is about debugging the UI process, we must create a new UI process to take its place. Because the debugged process will be suspended at some point, and suspending the UI process means freezing the UI of the image"
+	| defaultUIManager |
+	defaultUIManager := self defaultUIManager.
```
J'utilise le @ pour indiquer que le nom qui suit est le nom d'un argument. Quel est le problème?

StevenCostiou replied:
Ça n'est pas l'usage dans Pharo, c'est plus pytonique ou java. Les arguments restent avec le même nom.
C'est une question d'uniformité si on intègre à Pharo, c'est ni mal ni bien.

cdlm replied:
Pharo passera peut-être à une syntaxe de markup pour les commentaires mais pour l'instant y'en a pas, donc rester en texte simple pour l'instant.

StevenCostiou commented on src/NewTools-DebuggerSelector-Extensions/UIManager.extension.st:
```diff
@@ -0,0 +1,19 @@
+Extension { #name : #UIManager }
+
+{ #category : #'*NewTools-DebuggerSelector-Extensions' }
+UIManager >> handleDebugRequest: aDebugRequest [
+	"Provide a default handling for unhandled error. 
+	 We should never reach this code, because it should be overridden by subclasses. 
+	 (but we can't put 'self subclassResponsibility' , because it will cause infinite loop,
+	 in case if current ui manager having no override of this method).
+	 This method is only for documentation purposes."
+		
+	Smalltalk snapshot: false andQuit: true.
+]
```
Du coup est-ce que cela ne devrait pas être dans debugger system ?

StevenCostiou commented:
Il manque également les méthodes `signalDebuggerError:`de l'ancien debuggerselctor, je ne les vois pas.

StevenCostiou committed:
Debugger selection strategies packages

StevenCostiou committed:
Fixing tests setup

StevenCostiou committed:
Better debugger seleciton

StevenCostiou committed:
test comment

StevenCostiou committed:
Removing test that make no more sense

StevenCostiou committed:
test hanlde debugger errors

StevenCostiou committed:
handle debugger errors

StevenCostiou committed:
testing and fixing next debugger algorithm

StevenCostiou committed:
Dummy debugger improvements

StevenCostiou committed:
Tests try open session with debuggers and implementation

StevenCostiou committed:
Renaming handleDebuggerErrors to handlesDebuggerErrors

StevenCostiou committed:
Test debugger opening

StevenCostiou committed:
Removing old tests

StevenCostiou committed:
Refactoring tests, introducing DSSingleDebuggerSelectorStrategy

StevenCostiou committed:
Renaming, refactoring, retesting.

StevenCostiou committed:
Debugger selection strategy settings

StevenCostiou committed:
Typo

StevenCostiou committed:
Handle debugger errors settings

StevenCostiou committed:
Fixing tests and handle debugger errors initialization

dupriezt committed:
Merge pull request #72 from pharo-spec/Debuggers-Selection-Strategies

Debuggers selection strategies

StevenCostiou committed:
Handled for SingleDebuggerSelection tsrategy

StevenCostiou committed:
Adding an error inst var to hold the debugger error in case it happens

Commit:
Default debugger selection strategy is DSDebuggerSelector

Commit:
When DebugRequest creates a DebugSession, it also fills the #exception instance variable

Commit:
DebugRequest cannot be instanciated with #new

Commit:
defaultLabel_ -> defaultLabel

Commit:
formatting + cascading in #nullExceptionFor:named: to avoid temporary variable

Commit:
DebugRequest tests

Commit:
DebuggerSystem uses debugger selector strategy defined in setting belonging to DSDebuggerSelectionStrategy

Commit:
DebuggerSystem list existing debuggers

Commit:
Fix for UI process spawn when debug session is about UI process, to avoid image freeze when closing the window of the opened debugger

Commit:
Removed class-side API of DebuggerSystem + updated DebuggerSystemTest

Commit:
Remove @ in comments

Commit:
Warning>>#unhandledErrorAction submits a WarningDebugRequest

Commit:
categorise + comment

Commit:
More DebuggerSystem tests

Commit:
DummyDebuggerSelectionStrategy now subclass of DSDebuggerSelectionStrategy

Commit:
Remove DebuggerSelector_

Commit:
Copied debugger system error handling code from old DebuggerSelector class to DebuggerSystem class

Commit:
Factor code in DebuggerSystemTest

Commit:
When GT debugger opens, if it is one of its specialised variants (SUnit debugger or Bytecode debugger), it will call #spanNewSessionFrom: to get a specialised DebugSession.
This was the original behaviour, and removing it broke these specialised debuggers

StevenCostiou committed:
Merge pull request #73 from pharo-spec/Debuggers-Selection-Strategies

Debuggers selection strategies: handled and errors

StevenCostiou commented:
Integrated here: https://github.com/pharo-project/pharo/pull/7180
